add enum consts to work around an issue where prisma generates an enu…

gabriel-fishman-hike-medical opened the pull request:
…m and a type with the same way, which makes them incompatible with zod

basememara commented:
> I added this:
> 
> ```
> & {
>   [K in ProductType]: K;
> };
> ```
> 
> because it will force a compile-time error if a value of ProductType is missing, so it will force us to keep the enum in sync with Prisma. Do you think that's valuable to keep here? I'll change the rest to match the pattern.

I tried it and it worked! enforcing missing keys is exactly what we want but couldn't figure it out before, nice!

it would be great to change the others but probably best in a separate PR. but before enshrining it into convention, a couple of light comments:

1. Do we need to export `NotificationTypeEnum` or can we keep it private? just wondering if callers can use the Prisma enum version when needed instead? I feel the original purpose was to get the list but can use the Prisma enum otherwise.
2. Is casting `as [NotificationType, ...NotificationType[]]` needed or can it be `satisfies NotificationType[]`? the latter will make it a string union which makes it compatible in more places. also avoids `as` as a bonus. but wasn't sure if this was also a workaround to some restriction

gabriel-fishman-hike-medical commented:
> > I added this:
> > ```
> > & {
> >   [K in ProductType]: K;
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > because it will force a compile-time error if a value of ProductType is missing, so it will force us to keep the enum in sync with Prisma. Do you think that's valuable to keep here? I'll change the rest to match the pattern.
> 
> I tried it and it worked! enforcing missing keys is exactly what we want but couldn't figure it out before, nice!
> 
> it would be great to change the others but probably best in a separate PR. but before enshrining it into convention, a couple of light comments:
> 
> 1. Do we need to export `NotificationTypeEnum` or can we keep it private? just wondering if callers can use the Prisma enum version when needed instead? I feel the original purpose was to get the list but can use the Prisma enum otherwise.
> 2. Is casting `as [NotificationType, ...NotificationType[]]` needed or can it be `satisfies NotificationType[]`? the latter will make it a string union which makes it compatible in more places. also avoids `as` as a bonus. but wasn't sure if this was also a workaround to some restriction

1. Dropped the export.
2. I can drop the cast. I put that in to get zod to stop complaining, but I can just make that cast where it's used instead.
